File: codes/visual_basic_dotnet/array_find.vb
Imports System.Collections.Generic ' Dictionary, KeyValuePair

Module Program
    Function JsonStringify(ByVal Anything As Object, Optional ByVal Pretty As Boolean = False, Optional ByVal Indent As String = "    ") As String
        Dim IndentLevel As Integer = 0
        Dim JsonStringifyInner As Func(Of Object, String, String) = Function(ByVal AnythingInner As Object, ByVal IndentInner As String)
            If (AnythingInner Is Nothing) Then Return "null"
            If (TypeOf AnythingInner Is String) Then Return """" & AnythingInner & """"
            If (IsNumeric(AnythingInner) = True) Then Return CStr(AnythingInner).Replace(",", ".")
            If (TypeOf AnythingInner Is Boolean) Then Return CStr(AnythingInner)
            If (TypeOf AnythingInner Is List(Of Object)) Then
                If (AnythingInner.Count = 0) Then Return "[]"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "[" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "[")
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In AnythingInner
                    Result &= JsonStringifyInner(ArrayItem, IndentInner)
                    If ((ArrayItemIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ArrayItemIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "]", "]")
                Return Result
            End If
            If (TypeOf AnythingInner Is Dictionary(Of String, Object)) Then
                If (AnythingInner.Count = 0) Then Return "{}"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "{" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "{")
                Dim ObjectIterationIndex As Integer = 0
                For Each ObjectEntry As KeyValuePair(Of String, Object) In AnythingInner
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    Result &= """" & ObjectKey & """: " & JsonStringifyInner(ObjectValue, IndentInner)
                    If ((ObjectIterationIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ObjectIterationIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "}", "}")
                Return Result
            End If
            Return "null"
        End Function
        Return JsonStringifyInner(Anything, Indent)
    End Function

    Function ArrayFindV1(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Boolean), ByVal AnArray As List(Of Object)) As Object
        ' JavaScript-like Array.find() function
        Dim DataFound As Object = Nothing
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            Dim IsConditionMatch As Boolean = CallbackFunction(ArrayItem, ArrayItemIndex, AnArray)
            If (IsConditionMatch = true) Then
                DataFound = ArrayItem
                Exit For
            End If
            ArrayItemIndex += 1
        Next
        Return DataFound
    End Function

    Function ArrayFindV2(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Boolean), ByVal AnArray As List(Of Object)) As Object
        ' JavaScript-like Array.find() function
        Dim DataFound As Object = Nothing
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            If (CallbackFunction(ArrayItem, ArrayItemIndex, AnArray) = true) Then
                DataFound = ArrayItem
                Exit For
            End If
            ArrayItemIndex += 1
        Next
        Return DataFound
    End Function

    Function ArrayFindV3(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Boolean), ByVal AnArray As List(Of Object)) As Object
        ' JavaScript-like Array.find() function
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            Dim IsConditionMatch As Boolean = CallbackFunction(ArrayItem, ArrayItemIndex, AnArray)
            If (IsConditionMatch = true) Then Return ArrayItem
            ArrayItemIndex += 1
        Next
        Return Nothing
    End Function

    Function ArrayFindV4(ByVal CallbackFunction As Func(Of Object, Integer, List(Of Object), Boolean), ByVal AnArray As List(Of Object)) As Object
        ' JavaScript-like Array.find() function
        Dim ArrayItemIndex As Integer = 0
        For Each ArrayItem As Object In AnArray
            If (CallbackFunction(ArrayItem, ArrayItemIndex, AnArray) = true) Then Return ArrayItem
            ArrayItemIndex += 1
        Next
        Return Nothing
    End Function

    Sub Main(Args As String())
        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.find() in Visual Basic (.NET) List")

        Dim Numbers As List(Of Object) = New List(Of Object) From {12, 34, 27, 23, 65, 93, 36, 87, 4, 254}
        Console.WriteLine($"Numbers: {JsonStringify(Numbers)}")

        Dim EvenNumberFound As Object
        Dim OddNumberFound As Object

        Console.WriteLine("' using JavaScript-like Array.find() function ""ArrayFindV1""")

        EvenNumberFound = ArrayFindV1(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) ((Number Mod 2) = 0), Numbers)
        Console.WriteLine($"even number found: {EvenNumberFound}")
        ' even number found: 12

        OddNumberFound = ArrayFindV1(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) ((Number Mod 2) <> 0), Numbers)
        Console.WriteLine($"even number found: {OddNumberFound}")
        ' odd number found: 27

        Console.WriteLine("' using JavaScript-like Array.find() function ""ArrayFindV2""")

        EvenNumberFound = ArrayFindV2(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) ((Number Mod 2) = 0), Numbers)
        Console.WriteLine($"even number found: {EvenNumberFound}")
        ' even number found: 12

        OddNumberFound = ArrayFindV2(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) ((Number Mod 2) <> 0), Numbers)
        Console.WriteLine($"even number found: {OddNumberFound}")
        ' odd number found: 27

        Console.WriteLine("' using JavaScript-like Array.find() function ""ArrayFindV3""")

        EvenNumberFound = ArrayFindV3(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) ((Number Mod 2) = 0), Numbers)
        Console.WriteLine($"even number found: {EvenNumberFound}")
        ' even number found: 12

        OddNumberFound = ArrayFindV3(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) ((Number Mod 2) <> 0), Numbers)
        Console.WriteLine($"even number found: {OddNumberFound}")
        ' odd number found: 27

        Console.WriteLine("' using JavaScript-like Array.find() function ""ArrayFindV4""")

        EvenNumberFound = ArrayFindV4(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) ((Number Mod 2) = 0), Numbers)
        Console.WriteLine($"even number found: {EvenNumberFound}")
        ' even number found: 12

        OddNumberFound = ArrayFindV4(Function(ByVal Number As Integer, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) ((Number Mod 2) <> 0), Numbers)
        Console.WriteLine($"even number found: {OddNumberFound}")
        ' odd number found: 27

        Console.WriteLine("' using Visual Basic (.NET) Array.some() built-in method ""Array.Find()""")

        EvenNumberFound = Numbers.Find(Function(ByVal Number As Integer) ((Number Mod 2) = 0))
        Console.WriteLine($"even number found: {EvenNumberFound}")
        ' even number found: 12

        OddNumberFound = Numbers.Find(Function(ByVal Number As Integer) ((Number Mod 2) <> 0))
        Console.WriteLine($"even number found: {OddNumberFound}")
        ' odd number found: 27

        Console.WriteLine("' using Visual Basic (.NET) Array.some() built-in method ""Array.FirstOrDefault()""")

        EvenNumberFound = Numbers.FirstOrDefault(Function(ByVal Number As Integer) ((Number Mod 2) = 0))
        Console.WriteLine($"even number found: {EvenNumberFound}")
        ' even number found: 12

        OddNumberFound = Numbers.FirstOrDefault(Function(ByVal Number As Integer) ((Number Mod 2) <> 0))
        Console.WriteLine($"even number found: {OddNumberFound}")
        ' odd number found: 27

        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Array.find() in Visual Basic (.NET) List of Dictionaries")

        Dim Products As List(Of Object) = New List(Of Object) From {
            New Dictionary(Of String, Object) From {
                {"code", "pasta"},
                {"price", 321}
            },
            New Dictionary(Of String, Object) From {
                {"code", "bubble_gum"},
                {"price", 233}
            },
            New Dictionary(Of String, Object) From {
                {"code", "potato_chips"},
                {"price", 5}
            },
            New Dictionary(Of String, Object) From {
                {"code", "towel"},
                {"price", 499}
            }
        }
        Console.WriteLine($"Products: {JsonStringify(Products, Pretty:=True)}")

        Dim ProductToFind As String = "bubble_gum"
        Console.WriteLine($"product to find: {JsonStringify(ProductToFind)}")

        Dim ProductFound As Object

        Console.WriteLine("' using JavaScript-like Array.find() function ""ArrayFindV1""")

        ProductFound = ArrayFindV1(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("code") = ProductToFind), Products)
        Console.WriteLine($"product found: {JsonStringify(ProductFound, Pretty:=True)}")
        ' product found: {
        '     "code":"bubble_gum",
        '     "price": 233
        ' }

        Console.WriteLine("' using JavaScript-like Array.find() function ""ArrayFindV2""")

        ProductFound = ArrayFindV2(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("code") = ProductToFind), Products)
        Console.WriteLine($"product found: {JsonStringify(ProductFound, Pretty:=True)}")
        ' product found: {
        '     "code":"bubble_gum",
        '     "price": 233
        ' }

        Console.WriteLine("' using JavaScript-like Array.find() function ""ArrayFindV3""")

        ProductFound = ArrayFindV3(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("code") = ProductToFind), Products)
        Console.WriteLine($"product found: {JsonStringify(ProductFound, Pretty:=True)}")
        ' product found: {
        '     "code":"bubble_gum",
        '     "price": 233
        ' }

        Console.WriteLine("' using JavaScript-like Array.find() function ""ArrayFindV4""")

        ProductFound = ArrayFindV4(Function(ByVal Product As Object, ByVal ArrayItemIndex As Integer, ByVal AnArray As List(Of Object)) (Product("code") = ProductToFind), Products)
        Console.WriteLine($"product found: {JsonStringify(ProductFound, Pretty:=True)}")
        ' product found: {
        '     "code":"bubble_gum",
        '     "price": 233
        ' }

        Console.WriteLine("' using Visual Basic (.NET) Array.some() built-in method ""Array.Find()""")

        ProductFound = Products.Find(Function(ByVal Product As Object) (Product("code") = ProductToFind))
        Console.WriteLine($"product found: {JsonStringify(ProductFound, Pretty:=True)}")
        ' product found: {
        '     "code":"bubble_gum",
        '     "price": 233
        ' }

        Console.WriteLine("' using Visual Basic (.NET) Array.some() built-in method ""Array.FirstOrDefault()""")

        ProductFound = Products.FirstOrDefault(Function(ByVal Product As Object) (Product("code") = ProductToFind))
        Console.WriteLine($"product found: {JsonStringify(ProductFound, Pretty:=True)}")
        ' product found: {
        '     "code":"bubble_gum",
        '     "price": 233
        ' }
    End Sub
End Module

File: codes/go/array_find/array_find.go
package main

import (
	"encoding/json"
	"fmt"
	"reflect"
	"strings"
)

// type any interface{}
type array []any
type object map[string]any

func arrayReduce(callbackFunction func(any, any, int, array) any, anArray array, initialValue any) any {
    // JavaScript-like Array.reduce() function
    result := initialValue
    for arrayItemIndex, arrayItem := range anArray {
        result = callbackFunction(result, arrayItem, arrayItemIndex, anArray)
    }
    return result
}

func optionalChaining(anything any, objectPropertiesArray ...any) any {
    anythingType := reflect.TypeOf(anything).Kind()
    if (((anythingType != reflect.Map) && (anythingType != reflect.Slice)) || (len(objectPropertiesArray) == 0)) {
        return anything
    }
    return arrayReduce(func(currentResult any, currentItem any, _ int, _ array) any {
        if (currentResult == nil && (anythingType == reflect.Map) && (reflect.TypeOf(currentItem).Kind() == reflect.String)) {
            return anything.(object)[currentItem.(string)]
        }
        if (currentResult == nil && (anythingType == reflect.Slice) && (reflect.TypeOf(currentItem).Kind() == reflect.Int) && (currentItem.(int) >= 0) && (len(anything.(array)) > currentItem.(int))) {
            return anything.(array)[currentItem.(int)]
        }
        if (reflect.TypeOf(currentResult).Kind() == reflect.Map && (reflect.TypeOf(currentItem).Kind() == reflect.String)) {
            return currentResult.(object)[currentItem.(string)]
        }
        if ((reflect.TypeOf(currentResult).Kind() == reflect.Slice) && (reflect.TypeOf(currentItem).Kind() == reflect.Int) && (currentItem.(int) >= 0) && (len(currentResult.(array)) > currentItem.(int))) {
            return currentResult.(array)[currentItem.(int)]
        }
        return nil
    }, objectPropertiesArray, nil)
}

func ternary(trueCondition bool, callbackFunctionIfConditionTrue func() any, callbackFunctionIfConditionFalse func() any) any {
    if (trueCondition == true) {
        return callbackFunctionIfConditionTrue()
    }
    return callbackFunctionIfConditionFalse()
}

func jsonStringify(restArguments ...any) string {
	jsonStringifyInner := func(anythingInner any, prettyInner bool, indentInner string) string {
		if (prettyInner == true) {
			jsonStringifyInnerResult, err := json.MarshalIndent(anythingInner, "", indentInner)
			if (err == nil) {
				return string(jsonStringifyInnerResult)
			}
			return "null"
		}
        jsonStringifyInnerResult, err := json.Marshal(anythingInner)
		if (err == nil) {
			return strings.ReplaceAll(string(jsonStringifyInnerResult), ",", ", ")
		}
		return "null"
    }
    var anything any = restArguments[0]
	prettyDefault := false
	indentDefault := "    "
	var pretty any = prettyDefault
	var indent any = indentDefault
	if (len(restArguments) == 1) {
		return jsonStringifyInner(anything, prettyDefault, indentDefault)
	}
	var optionalArgument any = restArguments[1]
	if (optionalArgument == false) {
		return jsonStringifyInner(anything, prettyDefault, indentDefault)
	}
	if (reflect.TypeOf(optionalArgument).Kind() == reflect.Map) {
		pretty = optionalChaining(optionalArgument, "pretty")
		indent = optionalChaining(optionalArgument, "indent")
		pretty = ternary((pretty == nil), func() any { return prettyDefault }, func() any { return pretty })
		indent = ternary((indent == nil), func() any { return indentDefault }, func() any { return indent })
		return jsonStringifyInner(anything, pretty.(bool), indent.(string))
	}
	if (optionalArgument == true) {
		if (len(restArguments) >= 3) {
			var additionalArgument2 any = restArguments[2]
			if (reflect.TypeOf(additionalArgument2).Kind() == reflect.String) {
				indent = additionalArgument2
			}
		}
		pretty = optionalArgument
		return jsonStringifyInner(anything, pretty.(bool), indent.(string))
	}
	return jsonStringifyInner(anything, prettyDefault, indentDefault)
}

func sprint(restArguments ...any) {
    var newArray = []string{}
    for _, argument := range restArguments {
        argumentType := reflect.TypeOf(argument).Kind()
        if (argumentType == reflect.Slice && (len(argument.(array)) == 1)) {
            newArray = append(newArray, jsonStringify(argument.(array)[0]))
            continue
        }
        if (argumentType == reflect.String) {
			newArray = append(newArray, argument.(string))
            continue
        }
    }
    fmt.Println(strings.Join(newArray, ""))
}

func arrayFindV1(callbackFunction func(any, int, array) bool, anArray array) any {
    // JavaScript-like Array.find() function
    var dataFound any
    for arrayItemIndex, arrayItem := range anArray {
        isConditionMatch := callbackFunction(arrayItem, arrayItemIndex, anArray)
        if (isConditionMatch == true) {
            dataFound = arrayItem
            break
        }
    }
    return dataFound
}

func arrayFindV2(callbackFunction func(any, int, array) bool, anArray array) any {
    // JavaScript-like Array.find() function
    var dataFound any
    for arrayItemIndex, arrayItem := range anArray {
        if (callbackFunction(arrayItem, arrayItemIndex, anArray) == true) {
            dataFound = arrayItem
            break
        }
    }
    return dataFound
}

func arrayFindV3(callbackFunction func(any, int, array) bool, anArray array) any {
    // JavaScript-like Array.find() function
    for arrayItemIndex, arrayItem := range anArray {
        isConditionMatch := callbackFunction(arrayItem, arrayItemIndex, anArray)
        if (isConditionMatch == true) {
            return arrayItem
        }
    }
    return nil
}

func arrayFindV4(callbackFunction func(any, int, array) bool, anArray array) any {
    // JavaScript-like Array.find() function
    for arrayItemIndex, arrayItem := range anArray {
        if (callbackFunction(arrayItem, arrayItemIndex, anArray) == true) {
            return arrayItem
        }
    }
    return nil
}

func main() {
    fmt.Println("\n// JavaScript-like Array.find() in Go Slice")

    numbers := array{12, 34, 27, 23, 65, 93, 36, 87, 4, 254}
    sprint("numbers: ", jsonStringify(numbers))

    var evenNumberFound any
    var oddNumberFound any

    fmt.Println("// using JavaScript-like Array.find() function \"arrayFindV1\"")

    evenNumberFound = arrayFindV1(func(number any, _ int, _ array) bool {
        return ((number.(int) % 2) == 0)
    }, numbers)
    sprint("even number found: ", evenNumberFound)
    // even number found: 12

    oddNumberFound = arrayFindV1(func(number any, _ int, _ array) bool {
        return ((number.(int) % 2) != 0)
    }, numbers)
    sprint("odd number found: ", oddNumberFound)
    // odd number found: 27

    fmt.Println("// using JavaScript-like Array.find() function \"arrayFindV2\"")

    evenNumberFound = arrayFindV2(func(number any, _ int, _ array) bool {
        return ((number.(int) % 2) == 0)
    }, numbers)
    sprint("even number found: ", evenNumberFound)
    // even number found: 12

    oddNumberFound = arrayFindV2(func(number any, _ int, _ array) bool {
        return ((number.(int) % 2) != 0)
    }, numbers)
    sprint("odd number found: ", oddNumberFound)
    // odd number found: 27

    fmt.Println("// using JavaScript-like Array.find() function \"arrayFindV3\"")

    evenNumberFound = arrayFindV3(func(number any, _ int, _ array) bool {
        return ((number.(int) % 2) == 0)
    }, numbers)
    sprint("even number found: ", evenNumberFound)
    // even number found: 12

    oddNumberFound = arrayFindV3(func(number any, _ int, _ array) bool {
        return ((number.(int) % 2) != 0)
    }, numbers)
    sprint("odd number found: ", oddNumberFound)
    // odd number found: 27

    fmt.Println("// using JavaScript-like Array.find() function \"arrayFindV4\"")

    evenNumberFound = arrayFindV4(func(number any, _ int, _ array) bool {
        return ((number.(int) % 2) == 0)
    }, numbers)
    sprint("even number found: ", evenNumberFound)
    // even number found: 12

    oddNumberFound = arrayFindV4(func(number any, _ int, _ array) bool {
        return ((number.(int) % 2) != 0)
    }, numbers)
    sprint("odd number found: ", oddNumberFound)
    // odd number found: 27

    fmt.Println("\n// JavaScript-like Array.find() in Go Slice of maps")

    products := array{
        object{
            "code": "pasta",
            "price": 321,
        },
        object{
            "code": "bubble_gum",
            "price": 233,
        },
        object{
            "code": "potato_chips",
            "price": 5,
        },
        object{
            "code": "towel",
            "price": 499,
        },
    }
    sprint("products: ", jsonStringify(products, object{"pretty": true}))

    productToFind := "bubble_gum"
    sprint("product to find: ", productToFind)

    var productFound any

    fmt.Println("// using JavaScript-like Array.find() function \"arrayFindV1\"")

    productFound = arrayFindV1(func(product any, _ int, _ array) bool {
        return (product.(object)["code"] == productToFind)
    }, products)
    sprint("product found: ", jsonStringify(productFound, object{"pretty": true}))
    // product found: {
    //     "code":"bubble_gum",
    //     "price": 233
    // }

    fmt.Println("// using JavaScript-like Array.find() function \"arrayFindV2\"")

    productFound = arrayFindV2(func(product any, _ int, _ array) bool {
        return (product.(object)["code"] == productToFind)
    }, products)
    sprint("product found: ", jsonStringify(productFound, object{"pretty": true}))
    // product found: {
    //     "code":"bubble_gum",
    //     "price": 233
    // }

    fmt.Println("// using JavaScript-like Array.find() function \"arrayFindV3\"")

    productFound = arrayFindV3(func(product any, _ int, _ array) bool {
        return (product.(object)["code"] == productToFind)
    }, products)
    sprint("product found: ", jsonStringify(productFound, object{"pretty": true}))
    // product found: {
    //     "code":"bubble_gum",
    //     "price": 233
    // }

    fmt.Println("// using JavaScript-like Array.find() function \"arrayFindV4\"")

    productFound = arrayFindV4(func(product any, _ int, _ array) bool {
        return (product.(object)["code"] == productToFind)
    }, products)
    sprint("product found: ", jsonStringify(productFound, object{"pretty": true}))
    // product found: {
    //     "code":"bubble_gum",
    //     "price": 233
    // }
}
